api/src/main/java/io/druid/data/input/impl/JavaScriptParseSpec.java

```diff
@@ -64,11 +68,8 @@ public void verify(List<String> usedCols)
   @Override
   public Parser<String, Object> makeParser()
   {
-    if (!config.isEnabled()) {
-      throw new ISE("JavaScript is disabled");
-    }
-
-    return new JavaScriptParser(function);
+    parser = parser == null ? new JavaScriptParser(function) : parser;
```

Comment:
Is this thread-safe? I guess, maybe, although it might do extra work since not all threads will see the update at the same time. I'm also not sure if ParseSpecs need to be thread-safe; if it's an issue could you please double-check it?

Reply:
It looks that one ParseSpec (and one InputRowParser) is created per task and there's no thread-safety issue. One exception is DataSchema.getParser() which might create a new InputRowParser, but it's always called by a single thread. I additionally modified DataSchema to cache the created InputRowParser.

I recognize ThriftDeserialization (added in https://github.com/druid-io/druid/pull/3418) is designed for thread-safety, but am not sure how different threads can call it at the same time.